design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,132 @@
+```
+ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
+The type of a header backed column must be exactly `ARRAY<STRUCT<key STRING, value BYTES>>`. An error will be thrown if that is not the case.
+```
+ksql> CREATE STREAM B (my_headers ARRAY<BYTES> HEADERS);
+Error: Columns specified with the HEADERS keyword must be typed as ARRAY<STRUCT<key STRING, value BYTES>>.
+```
+Users can also mark a column with `HEADER(<key>)`, which will populate the column with the last header that matches the key. The data type must be `BYTES`
+```
+ksql> CREATE STREAM B (value BYTES HEADER(<key>));
+```
+If the key is not present in the header, then the value in the column will be `null`.
+
+Sources created with schema inferencing can have header columns, they would just have to be specified. For example, the statement `CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');` will create a stream with columns from Schema Registry, as well as one header column.
+
+### New byte-conversion functions
+
+There will also be new byte conversion functions to help decode header data:
+
+* `INT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 4 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getInt` method. Unsigned conversions will be manually calculated. If the result does not fit into an INTEGER then the method will throw an error.
+* `BIGINT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 8 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getLong` method. Unsigned conversions will be manually calculated. If the result does not fit into a BIGINT then the method will throw an error.
+* `DOUBLE_FROM_BYTES(bytes, byte_order='big')` - The input byte array must be exactly 8 bytes long. The conversion will by handled by `ByteBuffer`'s `getDouble` method.
+
+### Queries
+
+The headers columns will usable in any query like any other column, except `INSERT` query will be blocked. However, when persistent queries project header-backed columns, the header values are copied into the output row’s value (or key), not to headers in the output record.
+
+### Schema
+
+Header columns will fully be a part of the stream/table's schema. When running `SELECT *`, header columns will be included in the projection, and they will also appear when describing a source.
+
+## Design
+Columns in the logical schema will be represented by a new type of column `HeaderColumn` which will extend `Column` and contain an optional String named `key`. They will be in a new `HEADER` namespace, which functions identically to the `VALUE` namespace, except that the `StreamBuilder` will populate those columns with header values. 
```

Comment:
What is the optional `key` field in `HeaderColumn` about?

Reply:
Added some clarification